store/git.go

```diff
@@ -0,0 +1,159 @@
+}
+
+// NewGitReader returns a new git reader
+func NewGitReader(kubeClientset kubernetes.Interface, gitArtifact *v1alpha1.GitArtifact) (*GitArtifactReader, error) {
+	return &GitArtifactReader{
+		kubeClientset: kubeClientset,
+		artifact:      gitArtifact,
+	}, nil
+}
+
+func getSSHKeyAuth(privateSshKeyFile string) (transport.AuthMethod, error) {
+	var auth transport.AuthMethod
+	sshKey, err := ioutil.ReadFile(privateSshKeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read ssh key file. err: %+v", err)
+	}
+	signer, err := ssh.ParsePrivateKey([]byte(sshKey))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ssh key. err: %+v", err)
+	}
+	auth = &go_git_ssh.PublicKeys{User: "git", Signer: signer}
+	return auth, nil
+}
+
+func (g *GitArtifactReader) getGitAuth() (transport.AuthMethod, error) {
+	if g.artifact.Creds != nil {
+		// retrieve access key id and secret access key
+		username, err := GetSecrets(g.kubeClientset, g.artifact.Namespace, g.artifact.Creds.Username.Name, g.artifact.Creds.Username.Key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve username: err: %+v", err)
+		}
+		password, err := GetSecrets(g.kubeClientset, g.artifact.Namespace, g.artifact.Creds.Password.Name, g.artifact.Creds.Password.Key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve password: err: %+v", err)
+		}
+		return &http.BasicAuth{
+			Username: username,
+			Password: password,
+		}, err
+	}
+	if g.artifact.SSHKeyPath != "" {
+		return getSSHKeyAuth(g.artifact.SSHKeyPath)
+	}
+	return nil, nil
+}
+
+func (g *GitArtifactReader) readFromRepository(r *git.Repository) ([]byte, error) {
+	w, err := r.Worktree()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get working tree. err: %+v", err)
+	}
+
+	pullOpts := &git.PullOptions{
+		RemoteName:        "origin",
+		RecurseSubmodules: git.DefaultSubmoduleRecursionDepth,
+	}
+
+	auth, err := g.getGitAuth()
+	if err != nil {
+		return nil, err
+	}
+	if auth != nil {
+		pullOpts.Auth = auth
+	}
+
+	refName, err := g.getBranchOrTag(r, g.artifact.Branch, g.artifact.Tag)
+	if err != nil {
+		return nil, err
+	}
+	if refName != nil {
+		pullOpts.ReferenceName = *refName
+	}
+
+	if err := w.Pull(pullOpts); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, fmt.Errorf("failed to pull latest updates. err: %+v", err)
+	}
+
+	return ioutil.ReadFile(fmt.Sprintf("%s/%s", g.artifact.CloneDirectory, g.artifact.FilePath))
+}
+
+func (g *GitArtifactReader) getBranchOrTag(r *git.Repository, branch, tag string) (*plumbing.ReferenceName, error) {
+	if branch != "" {
+		branch, err := r.Branch(branch)
+		if err != nil {
+			return nil, fmt.Errorf("branch %s not found. err: %+v", branch, err)
+		}
+		return &branch.Merge, nil
+	}
+	if tag != "" {
+		tag, err := r.Tag(tag)
+		if err != nil {
+			return nil, fmt.Errorf("tag %s not found. err: %+v", tag, err)
+		}
+		refName := tag.Name()
+		return &refName, nil
+	}
+	return nil, nil
```

Comment:
do you really want to return `nil` error here? what happens if the `refName` is nil?

Reply:
updated